File: territory.py
#!/usr/bin/env python3
"""
territory.py — aggregate region metrics up to the TERRITORY grain and run the
same signal machinery on the territory-level market-share trajectory.

Why a separate layer: region_metrics is per region. The Sales Manager works at
the territory level, where total sales / growth / market-share deviation / FCST
attainment for the *whole* territory matter — and market share is NOT additive,
so it must be recomputed from summed own-vs-total franchise sales, not averaged.

Aggregation (per brand, MAT period, per year_month):
  own   = SUM(region sales_eur)
  total = SUM(region sales_eur * 100 / market_share)   # reconstruct franchise total
  territory MS        = own / total * 100
  mshare_deviation    = territory MS − national MS (same month)
  growth_py / growth_pp = from the territory MAT own-sales series
  growth_vs_fcst      = territory growth_py − national forecast growth
                        (forecast is national-only; recovered as a region's
                         growth_py_sales − growth_vs_fcst_eur)
The territory mshare_deviation series (asof-3..asof) feeds knowledge.signal_strength
and findings.classify, exactly like a region.
"""

from collections import defaultdict

import server
import knowledge
import findings


def _yms(con):
    return [r[0] for r in con.execute(
        "SELECT DISTINCT year_month FROM region_metrics ORDER BY year_month")]


def territory_metrics(brand, asof=None):
    defs = knowledge.load()
    con = server._conn()
    yms = _yms(con)
    if not asof:
        asof = yms[-1]
    i = yms.index(asof)

    rows = con.execute("""
        SELECT territory_name AS t, year_month AS ym,
               SUM(sales_eur) AS own, SUM(units) AS units,
               SUM(CASE WHEN market_share > 0 THEN sales_eur * 100.0 / market_share ELSE 0 END) AS total
        FROM region_metrics
        WHERE brand_name = ? AND period_type = 'MAT'
        GROUP BY territory_name, year_month
    """, (brand,)).fetchall()
    natrows = con.execute("""
        SELECT year_month AS ym, SUM(sales_eur) AS own, SUM(units) AS units,
               SUM(CASE WHEN market_share > 0 THEN sales_eur * 100.0 / market_share ELSE 0 END) AS total
        FROM region_metrics WHERE brand_name = ? AND period_type = 'MAT' GROUP BY year_month
    """, (brand,)).fetchall()
    nat = {r['ym']: {'own': r['own'], 'units': r['units'],
                     'ms': (r['own'] / r['total'] * 100 if r['total'] else 0)} for r in natrows}
    nat_ms = {ym: v['ms'] for ym, v in nat.items()}
    # national forecast growth (same for every region of the brand) at asof
    fr = con.execute("""
        SELECT growth_py_sales, growth_vs_fcst_eur FROM region_metrics
        WHERE brand_name = ? AND period_type = 'MAT' AND year_month = ?
          AND growth_py_sales IS NOT NULL AND growth_vs_fcst_eur IS NOT NULL LIMIT 1
    """, (brand, asof)).fetchone()
    nat_fcst_growth = (fr['growth_py_sales'] - fr['growth_vs_fcst_eur']) if fr else None
    rc = {r['t']: r['n'] for r in con.execute(
        "SELECT territory_name AS t, COUNT(DISTINCT region_name) AS n "
        "FROM region_metrics WHERE brand_name=? GROUP BY territory_name", (brand,))}
    # national TRUE attainment per period_type: actual ÷ forecast (level, not growth)
    fcst_m = {r['ym']: r['f'] for r in con.execute(
        "SELECT f.year_month AS ym, SUM(f.sales_eur) AS f FROM forecast f "
        "JOIN skus s USING(sku_id) JOIN brands b ON s.brand_id=b.brand_id "
        "WHERE b.brand_name=? GROUP BY f.year_month", (brand,))}
    nat_actual = {r['p']: r['s'] for r in con.execute(
        "SELECT period_type AS p, SUM(sales_eur) AS s FROM region_metrics "
        "WHERE brand_name=? AND year_month=? GROUP BY period_type", (brand, asof))}
    con.close()

    def _period_fcst(period):
        if period == 'Month':
            ms = [asof]
        elif period == 'RollQ':
            ms = yms[i - 2:i + 1]
        elif period == 'MAT':
            ms = yms[i - 11:i + 1]
        else:  # YTD
            ms = [m for m in yms if m[:4] == asof[:4] and m <= asof]
        vals = [fcst_m.get(m) for m in ms]
        return sum(vals) if ms and all(v is not None for v in vals) else None

    attainment = {}
    for period in ('Month', 'RollQ', 'YTD', 'MAT'):
        act, fc = nat_actual.get(period), _period_fcst(period)
        if act is not None and fc:
            attainment[period] = {'actual': round(act), 'fcst': round(fc),
                                  'pct': round(act / fc * 100, 1)}  # % of plan

    # National product summary (no peer level above it yet — franchise comes
    # later — so its market-share "signal" is the national share's own trajectory).
    nat_at = lambda off: nat.get(yms[i - off]) if 0 <= i - off < len(yms) else None
    ncur = nat.get(asof)
    national = None
    if ncur:
        npy, npp = nat_at(12), nat_at(1)
        ng_py = ((ncur['own'] - npy['own']) / npy['own'] * 100) if npy and npy['own'] else None
        ng_pp = ((ncur['own'] - npp['own']) / npp['own'] * 100) if npp and npp['own'] else None
        ngvf = (ng_py - nat_fcst_growth) if (ng_py is not None and nat_fcst_growth is not None) else None
        sh = [nat_at(off)['ms'] if nat_at(off) else None for off in (3, 2, 1, 0)]
        nsig = None
        if all(x is not None for x in sh):
            sh = [round(x, 2) for x in sh]
            nsig = knowledge.signal_strength(sh, direction='higher_is_better', defs=defs, kind='position')
        national = {
            'mat_sales': round(ncur['own']), 'units': round(ncur['units']),
            'growth_py': None if ng_py is None else round(ng_py, 1),
            'growth_pp': None if ng_pp is None else round(ng_pp, 1),
            'market_share': round(ncur['ms'], 1),
            'growth_vs_fcst': None if ngvf is None else round(ngvf, 1),
            'share_series': sh if all(x is not None for x in sh) else None,
            'signal': nsig,
        }

    byT = defaultdict(dict)
    for r in rows:
        ms = (r['own'] / r['total'] * 100) if r['total'] else 0
        byT[r['t']][r['ym']] = {'own': r['own'], 'units': r['units'], 'ms': ms,
                                'dev': ms - nat_ms.get(r['ym'], 0)}

    out = []
    for t, series in byT.items():
        if asof not in series:
            continue
        cur = series[asof]
        at = lambda off: series.get(yms[i - off]) if 0 <= i - off < len(yms) else None
        py, pp = at(12), at(1)
        g_py = ((cur['own'] - py['own']) / py['own'] * 100) if py and py['own'] else None
        g_pp = ((cur['own'] - pp['own']) / pp['own'] * 100) if pp and pp['own'] else None
        gvf = (g_py - nat_fcst_growth) if (g_py is not None and nat_fcst_growth is not None) else None
        devser = [at(off)['dev'] if at(off) else None for off in (3, 2, 1, 0)]
        sig = None
        if all(x is not None for x in devser):
            devser = [round(x, 2) for x in devser]
            sig = knowledge.signal_strength(devser, direction='higher_is_better',
                                            defs=defs, kind='position')
        out.append({
            'territory': t,
            'region_count': rc.get(t, 0),
            'mat_sales': round(cur['own']),
            'units': round(cur['units']),
            'growth_py': None if g_py is None else round(g_py, 1),
            'growth_pp': None if g_pp is None else round(g_pp, 1),
            'market_share': round(cur['ms'], 1),
            'mshare_deviation': round(cur['dev'], 2),
            'growth_vs_fcst': None if gvf is None else round(gvf, 1),
            'dev_series': devser if all(x is not None for x in devser) else None,
            'signal': sig,
        })
    out.sort(key=lambda x: -x['mat_sales'])
    if national is not None:
        national['attainment'] = attainment
    return {'brand': brand, 'asof': asof,
            'national_fcst_growth': None if nat_fcst_growth is None else round(nat_fcst_growth, 1),
            'national': national, 'territories': out}


if __name__ == "__main__":
    import sys
    d = territory_metrics(sys.argv[1] if len(sys.argv) > 1 else "Oncleris")
    print(f"{d['brand']} @ {d['asof']}  (national fcst growth {d['national_fcst_growth']}pp)\n")
    print(f"{'territory':16}{'MAT sales':>12}{'YoY%':>7}{'PoP%':>7}{'MS%':>7}{'dev':>7}{'vsFCST':>8}  {'signal':>10}")
    for r in d['territories']:
        sh = f"{r['signal']['shape']}/{r['signal']['net']:+.1f}" if r['signal'] else '—'
        print(f"{r['territory'][:16]:16}{r['mat_sales']:>12,}{r['growth_py']:>7}{r['growth_pp']:>7}"
              f"{r['market_share']:>7}{r['mshare_deviation']:>7}{r['growth_vs_fcst']:>8}  {sh:>10}")

File: studio/src/lib/territory.ts
// territory.ts — the TERRITORY-level readout for the report app's territory
// blocks (national summary + per-territory table).
//
// Reads two materialized grains directly (Tier 1-2, built by compute_metrics.py):
//   territory_metrics — per territory; the signal machinery runs on its
//                       mshare_deviation, signal resolved by ROLE
//                       (signal_roles.territory_position).
//   national_metrics  — per brand; the national summary's absolute columns
//                       (sales / share / growth / vs-forecast). Its share signal
//                       is the national share's own trajectory (no peer above
//                       brand, so its deviations are 0 by construction).
// Plan attainment (actual ÷ forecast level) is national-only and computed here
// from the forecast table.
import { metricsDb } from './db'
import {
  knowledgeDefs, signalTemplates, role, signalStrength, round, type Json,
} from './knowledge'

interface TRow {
  t: string
  ym: string
  sales_eur: number
  units: number
  market_share: number | null
  mshare_deviation: number | null
  growth_py_sales: number | null
  growth_pp_sales: number | null
  growth_vs_fcst_eur: number | null
  growth_deviation: number | null
}

// One national_metrics row (per brand × month × period); the brand's national position.
interface NRow {
  ym: string
  sales_eur: number
  units: number
  market_share: number | null
  growth_py_sales: number | null
  growth_pp_sales: number | null
  growth_vs_fcst_eur: number | null
}

const rnd1 = (v: number | null | undefined): number | null => (v == null ? null : round(v, 1))

export function territoryMetrics(brand: string, asofParam?: string): Json {
  const defs = knowledgeDefs()
  const sigTpl = signalTemplates(defs)[role(defs, 'territory_position')]
  const sigKind = sigTpl.strength_kind ?? 'position'
  const sigOffsets = [...sigTpl.lags].sort((a, b) => b - a)   // oldest -> now
  const db = metricsDb()

  const yms = (db.prepare('SELECT DISTINCT year_month FROM territory_metrics ORDER BY year_month').all() as { year_month: string }[])
    .map(r => r.year_month)
  const asof = asofParam || yms[yms.length - 1]
  const i = yms.indexOf(asof)

  const rows = db.prepare(`
    SELECT territory_name AS t, year_month AS ym, sales_eur, units,
           market_share, mshare_deviation, growth_py_sales, growth_pp_sales,
           growth_vs_fcst_eur, growth_deviation
    FROM territory_metrics
    WHERE brand_name = ? AND period_type = 'MAT'
  `).all(brand) as TRow[]
  const rc: Record<string, number> = {}
  for (const r of db.prepare('SELECT territory_name AS t, COUNT(*) AS n FROM regions GROUP BY territory_name').all() as { t: string; n: number }[]) {
    rc[r.t] = r.n
  }
  // national TRUE attainment per period_type: actual ÷ forecast (level, not growth)
  const fcstM: Record<string, number> = {}
  for (const r of db.prepare(`
    SELECT f.year_month AS ym, SUM(f.sales_eur) AS f FROM forecast f
    JOIN skus s USING(sku_id) JOIN brands b ON s.brand_id=b.brand_id
    WHERE b.brand_name=? GROUP BY f.year_month
  `).all(brand) as { ym: string; f: number }[]) {
    fcstM[r.ym] = r.f
  }
  const natActual: Record<string, number> = {}
  for (const r of db.prepare(`
    SELECT period_type AS p, SUM(sales_eur) AS s FROM territory_metrics
    WHERE brand_name=? AND year_month=? GROUP BY period_type
  `).all(brand, asof) as { p: string; s: number }[]) {
    natActual[r.p] = r.s
  }

  const byT: Record<string, Record<string, TRow>> = {}
  for (const r of rows) {
    ;(byT[r.t] = byT[r.t] ?? {})[r.ym] = r
  }
  // National row per month from the materialized national_metrics table (one row
  // per brand = the brand's national position). No longer back-derived from
  // territory sums — the absolute columns are computed correctly by the ETL.
  const natByYm: Record<string, NRow> = {}
  for (const r of db.prepare(`
    SELECT year_month AS ym, sales_eur, units, market_share,
           growth_py_sales, growth_pp_sales, growth_vs_fcst_eur
    FROM national_metrics
    WHERE brand_name = ? AND period_type = 'MAT'
  `).all(brand) as NRow[]) {
    natByYm[r.ym] = r
  }
  const ncur = natByYm[asof]
  const natFcstGrowth = ncur && ncur.growth_py_sales != null && ncur.growth_vs_fcst_eur != null
    ? ncur.growth_py_sales - ncur.growth_vs_fcst_eur
    : null

  const periodFcst = (period: string): number | null => {
    let ms: string[]
    if (period === 'Month') ms = [asof]
    else if (period === 'RollQ') ms = yms.slice(Math.max(0, i - 2), i + 1)
    else if (period === 'MAT') ms = yms.slice(Math.max(0, i - 11), i + 1)
    else ms = yms.filter(m => m.slice(0, 4) === asof.slice(0, 4) && m <= asof)   // YTD
    const vals = ms.map(m => fcstM[m])
    return ms.length && vals.every(v => v != null) ? vals.reduce((a, b) => a + b, 0) : null
  }

  const attainment: Record<string, Json> = {}
  for (const period of ['Month', 'RollQ', 'YTD', 'MAT']) {
    const act = natActual[period]
    const fc = periodFcst(period)
    if (act != null && fc) {
      attainment[period] = { actual: round(act, 0), fcst: round(fc, 0), pct: round(act / fc * 100, 1) }
    }
  }

  // National product summary (no peer level above it yet — franchise comes
  // later — so its market-share "signal" is the national share's own trajectory).
  const natAt = (off: number) => (i - off >= 0 && i - off < yms.length ? natByYm[yms[i - off]] : undefined)
  let national: Json | null = null
  if (ncur) {
    let sh: (number | null)[] = sigOffsets.map(off => natAt(off)?.market_share ?? null)
    let nsig: Json | null = null
    if (sh.every(x => x != null)) {
      sh = sh.map(x => round(x as number, 2))
      nsig = signalStrength(sh as number[], sigTpl.direction, defs, sigKind) as unknown as Json
    }
    national = {
      mat_sales: round(ncur.sales_eur, 0), units: round(ncur.units, 0),
      growth_py: rnd1(ncur.growth_py_sales),
      growth_pp: rnd1(ncur.growth_pp_sales),
      market_share: rnd1(ncur.market_share),
      growth_vs_fcst: rnd1(ncur.growth_vs_fcst_eur),
      share_series: sh.every(x => x != null) ? sh : null,
      signal: nsig,
      attainment,
    }
  }

  const out: Json[] = []
  for (const [t, series] of Object.entries(byT)) {
    const cur = series[asof]
    if (!cur) continue
    const at = (off: number) => (i - off >= 0 && i - off < yms.length ? series[yms[i - off]] : undefined)
    let devser: (number | null)[] = sigOffsets.map(off => at(off)?.mshare_deviation ?? null)
    let sig: Json | null = null
    if (devser.every(x => x != null)) {
      devser = devser.map(x => round(x as number, 2))
      sig = signalStrength(devser as number[], sigTpl.direction, defs, sigKind) as unknown as Json
    }
    out.push({
      territory: t,
      region_count: rc[t] ?? 0,
      mat_sales: round(cur.sales_eur, 0),
      units: round(cur.units, 0),
      growth_py: rnd1(cur.growth_py_sales),
      growth_pp: rnd1(cur.growth_pp_sales),
      market_share: rnd1(cur.market_share),
      mshare_deviation: cur.mshare_deviation == null ? null : round(cur.mshare_deviation, 2),
      growth_vs_fcst: rnd1(cur.growth_vs_fcst_eur),
      dev_series: devser.every(x => x != null) ? devser : null,
      signal: sig,
    })
  }
  out.sort((a, b) => (b.mat_sales as number) - (a.mat_sales as number))
  return {
    brand, asof,
    national_fcst_growth: natFcstGrowth == null ? null : round(natFcstGrowth, 1),
    national, territories: out,
  }
}
